Overlay/Overlay.js

```diff
@@ -0,0 +1,223 @@
+import { disableBodyScroll, enableBodyScroll } from '../node_modules/body-scroll-lock/lib/bodyScrollLock.es6.js';
+import overlayEvents from './overlayEvents.js';
+import getAndValidateAttribute from '../shared/getAndValidateAttribute.mjs';
+
+/**
+ * Overlay that is opened/closed by open/closeoverlay events. Optionally closes on esc or
+ * click outside and always locks background (prevents scrolling).
+ */
+export default class Overlay extends window.HTMLElement {
+
+    /**
+     * Holds open state; should only be accessed through getter isOpen and modified through
+     * open/close.
+     * @private
+     */
+    internalIsOpen = false;
+
+    /* global window */
+
+    constructor() {
+        super();
+        // Store bound open function; we need the reference to de-register when component is
+        // disconnected from dom
+        this.boundHandleOpenEvent = this.handleOpenEvent.bind(this);
+        this.boundHandleCloseEvent = this.handleCloseEvent.bind(this);
+        this.boundHandleEscEvent = this.handleEscEvent.bind(this);
+        this.boundHandleClickOutsideEvent = this.handleClickOutsideEvent.bind(this);
+    }
+
+    connectedCallback() {
+        window.addEventListener(overlayEvents.get('openOverlay'), this.boundHandleOpenEvent);
+        window.addEventListener(overlayEvents.get('closeOverlay'), this.boundHandleCloseEvent);
+    }
+
+    disconnectedCallback() {
+        window.removeEventListener(overlayEvents.get('openOverlay'), this.boundHandleOpenEvent);
+        window.removeEventListener(overlayEvents.get('closeOverlay'), this.boundHandleCloseEvent);
+    }
+
+    /**
+     * @private
+     */
+    setupEscListener() {
+        window.addEventListener('keydown', this.boundHandleEscEvent);
+    }
+
+    /**
+     * @private
+     */
+    teardownEscListener() {
+        window.removeEventListener('keydown', this.boundHandleEscEvent);
+    }
+
+    /**
+     * @private
+     */
+    setupClickOutsideListener() {
+        window.addEventListener('click', this.boundHandleClickOutsideEvent);
+    }
+
+    /**
+     * @private
+     */
+    teardownClickOutsideListener() {
+        window.removeEventListener('click', this.boundHandleClickOutsideEvent);
+    }
+
+    /**
+     * @private
+     */
+    handleCloseEvent(event) {
+        if (event.detail && event.detail.overlayName === this.getAndValidateOverlayName()) {
+            this.close();
+        }
+    }
+
+    /**
+     * @private
+     */
+    handleOpenEvent(event) {
+        if (event.detail && event.detail.overlayName === this.getAndValidateOverlayName()) {
+            this.open();
+        }
+    }
+
+    /**
+     * Returns current open state of the overlay
+     * @return {boolean}
+     */
+    get isOpen() {
+        return this.internalIsOpen;
```

Comment:
maybe: ```return this.classList.contains(this.visibleClassName);``` so you would not need the internalIsOpen state?

Reply:
There are three reasons for using the internal `isOpen` state:
- We don't have to access the DOM which is not very performant.
- We update the DOM asynchronously (through `requestAnimationFrame`) – the DOM does therefore never contain the **most current** state, it's max. ~16ms late. If we open/hide the overlay in quick successtion, the DOM and the actual state will be out of sync which is the main issue here. Using the internal `isOpen` state allows us to update the DOM reactively: We request the `animationFrame`; as soon as the `animationFrame` happens, it takes the most recent state from the model and updates the DOM accordingly. That's really the best practice to do any DOM updates.
- And generally I'm not sure (or doubt) that the DOM is the right place to store state, especially in classes which mainly have styling purposes.